### PR TITLE
Added global commands which allow a user to easily learn the destination of a link

### LINKED ISSUE #14535: Add ability to report destination URL of a link
### Is your feature request related to a problem? Please describe.

Links on web pages and in emails are often denoted with human-friendly text.  Particularly in an era where users need to be aware of and careful to avoid malicious sites and phishing attempts, it is important to check suspicious URLs before activating them.

### Describe the solution you'd like

I envisage the workflow as:
1. Moving focus to a link in an email or on a web page
2. Press a keystroke.
3. NVDA reports something like "Destination is <url> <title>", eg "Destination is https://www.nvaccess.org/download/ NV Access | Download NVDA"

### Describe alternatives you've considered

It used to be possible to query the status bar of the browser for information like this but this is generally not possible now.

The main workaround I can think of is:
1) Move to the link
2) Press the applications key to open the context menu
3) Activate the "Copy link", "Copy link address" or similar item.
4) Press NVDA+c to read the clipboard.

This is an additional step, and also overwrites the clipboard which isn't always ideal.

### PR DESCRIPTION
**Pull request by XLTechie**
### Link to issue number:

Closes #14535

### Summary of the issue:

@Qchristensen pointed out, that in this age of phishing and other daily digital security threats effecting average users across the globe, the ability to determine where a link on the web, in an email, or anywhere else, is actually going to take you, is of increasing importance.

NVDA has no current way to do that in most/all circumstances.

### Description of user facing changes

Added the following global commands:
* NVDA+k (all layouts): Report the destination URL of the link in the navigator object. Pressed once, will speak/braille the link, or other element containing a link, represented by the navigator object. Pressed twice, will show the link text in a window with the link name as its title, which can be reviewed more carefully.
* Unassigned by default: Report the destination URL of the link in the navigator object in a window. Will skip directly to showing the browseable message, as above. This second option is imagined as an alternative to which braille users might prefer to remap NVDA+k.

The NVDA+k key was chosen because of its relatability to the browse mode K command.

I also added the `NVDA+k` key and description, to the "**Reporting location and other information**" section of the user guide. I did not add information about the unmapped command, or a broader description of the feature in general, because I was not sure of an appropriate section for it, though it may be worth adding a "Miscellaneous Features" section at some point.

### Description of development approach

* Added two new scripts to `globalCommands.py`. The first will output the link, if the navigator object is or contains a link (such as a graphic with a link), as a `ui.message` if called once, or as a `ui.browseableMessage` if called twice.
* If the role is not LINK, or the states do not contain the LINKED state, it speaks/brailles "Not a link".
* The second script just calls the first script, with a flag directing it to ignore whether the key was pressed twice, and proceed directly to displaying in a `ui.browseableMessage`.

### Testing strategy:

* Manually tested links, and things that aren't links, in Firefox and Chromium web browsers.
* Tested that links are correctly spoken in MS Outlook browse mode, in either the reader or composer.
* Tested that links are correctly spoken when reading a message in MS Outlook focus mode.

### Known issues with pull request:

I can only make this work in MS Word browse mode, if I set UIA to always.
In focus mode, neither UIA or DOM indicate links properly.

While I consider this an issue to be solved, I think the fact that the current implementation works in most other places, warrants submitting the PR now.

I am not equipped to test this with braille or for low vision.

### Change log entries:
New features
* New global command: Report link destination (NVDA+k). Pressed once will speak/braille the destination of the link that is in the navigator object. Pressing twice will show it in a window, for more detailed review.
* New unmapped global command (Tools category): Report link destination in a window. Same as pressing NVDA+k twice, but may be more useful for braille users.

### Code Review Checklist:

- [X] Pull Request description:
  - description is up to date
  - change log entries
- [X] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [X] API is compatible with existing add-ons.
- [X] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [X] Security precautions taken.

### CODE REVIEW

**Comment by josephsl** [comment]
Hi, this will interfere with Place Markers add-on by @NVDAEs. Thanks.

**Review by CyrilleB79 — COMMENTED** [review]
LGTM globally. Just providing some minor suggestions.

Also, I think that this script is needed by a lot of people. Thus, I support the fact that you have attributed a gesture to the script (instead of leaving it unassigned). I think that there is also an equivalent script in recent versions of Jaws.

**Comment by XLTechie** [comment]
@josephsl wrote:
> Hi, this will interfere with Place Markers add-on by @NVDAEs. Thanks.

While I personally don't think add-ons should take one or two-modifier 
alphabetic commands--since NVDA is likely to need them in the future--I'm 
certainly open to alternative key ideas.

File: source/globalCommands.py [hunk 1]
```diff
@@ -3628,6 +3628,50 @@ def script_reloadPlugins(self, gesture):
...
+		@type forceBrowseable: bool
+		"""
+		obj = api.getNavigatorObject()
+		if (
+			obj.role == controlTypes.role.Role.LINK  # If it's a link, or
+			or controlTypes.state.State.LINKED in obj.states  # if it isn't a link but contains one
+		):
+			if (
+				forceBrowseable  # If a browseable message is preferred unconditionally, or
+				or scriptHandler.getLastScriptRepeatCount() > 0  # if pressed more than once
```
[the remark this answers is not in the document]
Comment (XLTechie) [reply source/globalCommands.py:3652]: @CyrilleB79  I am unclear if your comment relates to logic or behavior.

I wrote the code like that, so we could do the test for forceBrowseable, before running the pressed more than once check. That way, the check for number of presses doesn't get run if it isn't needed.
To do that with as little if-thening as possible, I had to check the two-press state before the one-press state.

If you would find it more correct, I can change the "> 0" to "== 1", which aught to produce the same effect, but would leave the three-press case undefined.

As it stands, it doesn't react to three presses. That is, if you press three times, it has already opened the browseableMessage, and the third press is registered as a new first press, which says "not a link". Is that not the expected behavior?

Your other comments are fixed in latest commit - thanks.

**Comment by josephsl** [comment]
Hi, one possible justification for assigning NVDA+D instead: 1800048 - Fetching IA2/ATK relations on links in the WHATWG HTML spec is slow (mozilla.org) <https://bugzilla.mozilla.org/show_bug.cgi?id=1800048> 

 

If links are treated as annotations (some might argue that), then I think treating hyperlink ref would be a good use of NVDA+D unless people think otherwise. NVDA+K makes sense from consistency perspective. Thanks.

**Comment by XLTechie** [comment]
> Hi, one possible justification for assigning NVDA+D instead

I may have missed something, if the existing script on NVDA+d is going to be 
removed, I suppose we could.

File: source/globalCommands.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (CyrilleB79) [reply source/globalCommands.py:3654]: > As it stands, it doesn't react to three presses. That is, if you press three times, it has already opened the browseableMessage, and the third press is registered as a new first press, which says "not a link". Is that not the expected behavior?

Yes, it is the expected behaviour. However on my system, when I triple-press, I get 2 browseable messages opened. Probably at the moment I perform the third press, the focus has not yet changed.

One very little off-topic question to native English spea

**Comment by Adriani90** [comment]
I think the discussion about key commands is rather one about prioritization. In my opinion the key commands that are becomming the core of NVDA should be in favor compared to addons. Thus, actually the commands of the addons could be changed instead. This would also bring the benefit of the addon developers being incentivized to update their addons in order to work properly.

**Comment by nvdaes** [comment]
Thanks for letting me know about possible conflict with placeMarkers. In case it"s needed, I"ll use a different command in the add-on.

**Comment by XLTechie** [comment]
Thank you @nvdaes. Would @Qchristensen or @seanbudd care to comment on whether NVDA+k is the preferred key for this, or if something else would be more desirable?

**Comment by ChrisDuffley** [comment]
Hello,

Please keep in mind that NVDA+K is a conflicting command for Braille Extender's Reload Driver function. Consider an alternate for either this proposed change or, if one of the maintainers for it, the mentioned add-on.

Thanks.
Chris

On Jan 25, 2023, at 21:18, Luke Davis ***@***.***> wrote:


Thank you @nvdaes <https://github.com/nvdaes>. Would @Qchristensen <https://github.com/Qchristensen> or @seanbudd <https://github.com/seanbudd> care to comment on whether NVDA+k is the preferred key for this, or if something else would be more desirable?

**Comment by XLTechie** [comment]
@CyrilleB79 I was only able once to replicate the behavior you described. Try this next build when it's done please, and let me know if the double window problem persists.

Thanks for your testing.

**Comment by CyrilleB79** [comment]
> @CyrilleB79 I was only able once to replicate the behavior you described. Try this next build when it's done please, and let me know if the double window problem persists.
> 
> Thanks for your testing.

The double window problem is not present anymore.

The logic of the code remains a bit strange, i.e. for a triple press, the function reports again the URL instead of doing nothing. However, from the UX point of view, it has no impact since the reported URL is not heard due to the browseable message opening and generating a focus event. You may want to do the code clearer, replacing the "`else`" by: "`elif scriptHandler.getLastScriptRepeatCount() == 0:`"

**Comment by XLTechie** [comment]
@CyrilleB79 wrote:
> The logic of the code remains a bit strange, i.e. for a triple press, the function reports again the URL instead of doing nothing. However, from the UX
> point of view, it has no impact since the reported URL is not heard due to the browseable message opening and generating a focus event. You may want to do
> the code clearer, replacing the "else" by: "elif scriptHandler.getLastScriptRepeatCount() == 0:"

Not how I did it, but I believe I have achieved what you want.
Check latest build (running now).

As an aside: you started to ask something to native English speakers, in one of 
your PR comments. Your question appeared cut-off. I tried to email you to 
continue that question, but your email system rejected my message, does not like 
my mail provider for some reason.

**Comment by CyrilleB79** [comment]
> @CyrilleB79 wrote:
> > The logic of the code remains a bit strange, i.e. for a triple press, the function reports again the URL instead of doing nothing. However, from the UX point of view, it has no impact since the reported URL is not heard due to the browseable message opening and generating a focus event. You may want to do the code clearer, replacing the "else" by: "elif scriptHandler.getLastScriptRepeatCount() == 0:"
> Not how I did it, but I believe I have achieved what you want. Check latest build (running now).

The code is clearer now, thanks. And of course it is still working correctly.

> As an aside: you started to ask something to native English speakers, in one of your PR comments. Your question appeared cut-off. I tried to email you to continue that question, but your email system rejected my message, does not like my mail provider for some reason.

It was a piece of message left over unintentionally. I had then edited my message to remove it. I was just concerned by the spelling of "browseable" but after some research, it seems that "browseable" is just an alternate valid spelling for "browsable".

**Comment by XLTechie** [comment]
@CyrilleB79 wrote:
> It was a piece of message left over unintentionally. I had then edited my message to remove it. I was just concerned by the spelling of "browseable" but
> after some research, it seems that "browseable" is just an alternate valid spelling for "browsable".

Apparently. I went through the same thing when I first found the function 
ui.browseableMessage, some years ago. That's not how we would do it in U.S. 
English.

Therefore while I have never liked that spelling, evidently that's how we spell 
it in NVDA. That's the only reason I have spelled it that way here--I guess it 
is preferred . . .

**Comment by Qchristensen** [comment]
Re the keystroke.  My first thought was NVDA+d, thinking of the old "long description" use.  Now that is used for aria-describedby, it could do double duty, although we'd need to work out what would happen if something had both aria-describedby AND a link.  Otherwise, NVDA+k is an option, although as noted, that conflicts with several add-ons.  Increasingly I think it will be hard to find new NVDA+one key commands which don't conflict with add-ons.

So, in short, I don't have a firm "It absolutely must be this keystroke" opinion, but just to throw one more in to the mix:  NVDA+numpad delete / NVDA+delete.  "Reports information about the location of the text or object at the position of system caret. "  Currently, it tells you the location on screen, which seems like a very rare use case, compared to wanting to know where a link will take you.

**Comment by michaelDCurran** [comment]
My preference would probably be NVDA+k. It relates to k (link) in quicknav.
I do acknowledge that this gesture is already taken by certain add-ons, 
but in that case, the add-on should win anyway - NVDA won't steal the 
existing gesture from the add-on, though does depend on exactly how the 
add-on was implemented.  However, the add-on author / user could change 
the gesture if they wanted to get to this new script.
Putting it in to either NvDA+d or NVDA+numpadDelete is possible, but 
could complicate things as to when and when not certain information is 
reported.

**Comment by cary-rowen** [comment]
I still agree with nvda +K, @nvdaes expressed willingness to change this gesture in PlaceMarkers, and other add-ons that may exist may be requested by users to change the gesture

**Review by michaelDCurran — APPROVED** [review]

**Review by Qchristensen — APPROVED** [review]
Looks good and will be very useful, thanks!

**Comment by Qchristensen** [comment]
Just trying this alpha build, it all works great.  When you press the keystroke twice, the window displayed is nearly the height of the screen, and about half the width of the screen.  The URL - which on the link I tested (the download link for the alpha build) is wider than the window - continues off the right edge of the window.  The window has a scrollbar to accommodate this, however it seems odd to have such a high window when the link, which doesn't wrap, can only ever be one line high.  It might be worth having the link wrap so that it is all visible (and ideally having the window fit to size so everything is neatly displayed on screen in a window which isn't overly large for the text in it.

Screenshot of how it appears on my system:
https://www.dropbox.com/s/snl03ub8nl03yu1/links.jpg?dl=0

And for those it helps, here is a (very rough) mockup of how I'd maybe have the window resized:

https://www.dropbox.com/s/63h0rlcikb3fx4m/links2.jpg?dl=0

(Happy to write up a new issue if that is useful - the functionality as it stands is very much more useful than not having it, so I would not propose reverting it, I'm just suggesting this as an improvement for the dialog when the keystroke is pressed twice).

**Comment by XLTechie** [comment]
@Qchristensen
> It might be worth having the link wrap so that it is all visible (and ideally having the window fit to size so everything is
> neatly displayed on screen in a window which isn't overly large for the text in it.

I agree with the second point--I foolishly didn't have anyone sighted review 
this, mistakenly assuming that ui.browseableMessage had common sense auto 
fitting.

I'll see what can be done in a followup PR.

Regarding your first point, however, I would hesitate to do a line break, 
because it would inhibit the user's ability to select all and copy the link, if 
desired.
While that is not the purpose of this feature, and the browser's copy 
destination feature should be what is used, we can't prevent that usage, and so 
it should behave as the user expects.

The best I *think* we can do, is to have it stylistically break every X 
characters (40?), but not break when selected and copied.

Something like described here: 
https://stackoverflow.com/questions/28376684/breaking-long-strings-visually-into-chunks-using-css-but-retaining-original-for

That will not resize with the Window, so presumably the window could be shrunk 
to the point where it would again be more narrow than the URL, but it might 
solve the immediate "bad look" problem.

Thoughts?

**Comment by Qchristensen** [comment]
I'm just trying to think of an equivalent dialog - the speech viewer dialog works how I had in mind - text wraps as needed to fit within the window - but if you copy the text to say Notepad, the text only breaks at a proper line break (or end of an utterance at least) - not where it wrapped in the speech viewer.

I agree re the need to keep the link whole for the purpose of copying it - while it is not necessarily the most direct way to copy the link originally, if you've opened this window to review it, the next logical action may be to want to open the URL if you are happy, and one common way to do that from here would likely be to copy it in order to open it in a new browser window

**Comment by XLTechie** [comment]
Note: the Outlook test above no longer passes. At least in Outlook 2019, I can not cause links to be read in either browse or focus mode, while either composing or reading.
